File: src/DocumentExport.Excel/MacroExecutor.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Runtime.InteropServices;
using Microsoft.Office.Interop.Excel;

namespace DocumentExport.Excel {

	/// <summary>
	/// Excelのマクロを実行する機能を提供します。
	/// </summary>
	internal class MacroExecutor {

		public void Execute(string filePath, string macroName) {
			Application application = null;
			Workbooks workbooks = null;
			Workbook workbook = null;

			try {
				application = new ApplicationClass();
				workbooks = application.Workbooks;

				workbook = workbooks.Open(
					filePath, Type.Missing, Type.Missing, Type.Missing, Type.Missing
					, Type.Missing, Type.Missing, Type.Missing, Type.Missing, Type.Missing
					, Type.Missing, Type.Missing, Type.Missing, Type.Missing, Type.Missing);

				application.Run(
					macroName, Type.Missing, Type.Missing, Type.Missing, Type.Missing
					, Type.Missing, Type.Missing, Type.Missing, Type.Missing, Type.Missing
					, Type.Missing, Type.Missing, Type.Missing, Type.Missing, Type.Missing
					, Type.Missing, Type.Missing, Type.Missing, Type.Missing, Type.Missing
					, Type.Missing, Type.Missing, Type.Missing, Type.Missing, Type.Missing
					, Type.Missing, Type.Missing, Type.Missing, Type.Missing, Type.Missing
					, Type.Missing);
			} finally {
				if (workbook != null) {
					try {
						workbook.Close(true, Type.Missing, Type.Missing); 
					} catch {
					}
				}

				if (application != null) {
					try {
						application.Quit();
					} catch {
					}
				}

				ReleaseComObjects(workbook, workbooks, application);
			}
		}

		/// <summary>
		/// 
		/// </summary>
		/// <param name="objects"></param>
		private void ReleaseComObjects(params object[] objects) {
			try {
				for (int i = 0; i < objects.Length; i++) {
					if (objects[i] == null)
						continue;
					if (Marshal.IsComObject(objects[i]) == false)
						continue;

					Marshal.ReleaseComObject(objects[i]);
					objects[i] = null;					
				}
			} catch {
			}
		}
	}
}

File: src/ClassLibrary1/MacroExecutor.vb
Imports System
Imports System.Collections.Generic
Imports System.Runtime.InteropServices
Imports Microsoft.Office.Interop.Excel

Public Class MacroExecutor
	Public Sub Execute(ByVal filePath As String, ByVal macroName As String)
		Dim application As Application = Nothing
		Dim workbooks As Workbooks = Nothing
		Dim workbook As Workbook = Nothing

		Try
			application = New ApplicationClass
			workbooks = application.Workbooks
			workbook = workbooks.Open(filePath)
			application.Run(macroName)
		Finally
			If workbook IsNot Nothing Then
				Try
					workbook.Close(True)
				Catch
				End Try
			End If

			If application IsNot Nothing Then
				Try
					application.Quit()
				Catch
				End Try
			End If

			ReleaseComObjects(workbook, workbooks, application)
		End Try

	End Sub

	Private Sub ReleaseComObjects(ByVal ParamArray objects As Object())
		Try
			For i As Integer = 0 To objects.Length
				If objects(i) Is Nothing Then
					Continue For
				End If

				If Marshal.IsComObject(objects(i)) = False Then
					Continue For
				End If

				Marshal.ReleaseComObject(objects(i))
			Next
		Catch
		End Try
	End Sub
End Class
